pkg/process/monitor/process_monitor.go

```diff
@@ -9,87 +9,68 @@
-	processMonitor *ProcessMonitor
+	processMonitor = &ProcessMonitor{
+		// Must initialize the sets, as we can register callbacks prior to calling Initialize.
+		processExecCallbacks: make(map[*ProcessCallback]struct{}, 0),
+		processExitCallbacks: make(map[*ProcessCallback]struct{}, 0),
+	}
 )
 
-// ProcessMonitor will subscribe to the netlink process events like Exec, Exit
-// and call the subscribed callbacks
-// Initialize() will scan the current process and will call the subscribed callbacks.
-//
-// callbacks will be executed in parallel via a pool of goroutines (runtime.NumCPU())
-// callbackRunner is callbacks queue. The queue size is set by processMonitorMaxEvents
-//
-// Multiple team can use the same ProcessMonitor,
-// the callers need to guarantee calling each Initialize() Stop() one single time
-// this maintains an internal reference counter
-//
+// ProcessMonitor uses netlink process events like Exec and Exit and activate the registered callbacks for the relevant
+// events.
 // ProcessMonitor require root or CAP_NET_ADMIN capabilities
 type ProcessMonitor struct {
-	m        sync.Mutex
-	wg       sync.WaitGroup
-	refcount int
-
-	isInitialized bool
-
-	// chan push done by vishvananda/netlink library
-	events chan netlink.ProcEvent
-	done   chan struct{}
-	errors chan error
+	initOnce sync.Once
+	// A wait group to give the Stop method an option to wait until the main event loop finished its teardown.
+	processMonitorWG sync.WaitGroup
+	// A wait group to give us the option to wait until all callback runners have finished.
+	callbackRunnersWG sync.WaitGroup
+	// An atomic counter to know how much instances do we have in any given time. Used to ensure when to clean up all
+	// resources.
+	refcount atomic.Int32
+	// A channel to mark the main routines to halt.
+	done chan struct{}
+
+	// netlink channels for process event monitor.
+	netlinkEventsChannel chan netlink.ProcEvent
+	netlinkDoneChannel   chan struct{}
+	netlinkErrorsChannel chan error
 
 	// callback registration and parallel execution management
-	procEventCallbacks map[ProcessEventType][]*ProcessCallback
-	runningPids        map[uint32]interface{}
-	callbackRunner     chan func()
+	processExecCallbacksMutex sync.RWMutex
+	processExecCallbacks      map[*ProcessCallback]struct{}
+	processExitCallbacksMutex sync.RWMutex
+	processExitCallbacks      map[*ProcessCallback]struct{}
+	callbackRunner            chan func()
 
 	// monitor stats
-	eventCount int
-	execCount  int
-	exitCount  int
-}
-
-type ProcessEventType int
-
-const (
-	EXEC ProcessEventType = iota
-	EXIT
-)
-
-type ProcessMetadataField int
-
-const (
-	ANY ProcessMetadataField = iota
-	NAME
-)
-
-type metadataName struct {
-	Name string
+	eventCount atomic.Uint32
+	execCount  atomic.Uint32
+	exitCount  atomic.Uint32
 }
 
 type ProcessCallback struct {
-	Event    ProcessEventType
-	Metadata ProcessMetadataField
-	Regex    *regexp.Regexp
-	Callback func(pid uint32)
+	Callback func(pid int)
```

Comment:
as ProcessCallback has been stripped from all metadata filter/feature and this design push the filter logic in the callback
why not set a simple type, this would simplify the caller code, or directly func(pid int) type in SubscribeExec/Exit()
`type ProcessCallback func(pid int)`

Reply:
that's not a valid key for map

Reply:
map[*func(int)] is valid, as we register different function for each subscriber
https://go.dev/play/p/ezPLMwjDEy2